configure.ac

```
@@ -444,10 +444,20 @@ if test "x$with_dl" != "xno"; then
 fi
 
 ## check for atomics library before anything that might need it
-# AC_SEARCH_LIBS pollutes LIBS
 SQUID_STATE_SAVE(LIBATOMIC)
-AC_SEARCH_LIBS([__atomic_load_8],[atomic],[
-  test "$ac_res" = "none required" || ATOMICLIB=$ac_res],[])
+AC_MSG_CHECKING(library containing __atomic_is_lock_free)
+LIBS="-latomic"
+AC_LINK_IFELSE(
+  [AC_LANG_SOURCE([[
+    int main(void) {
+        char lockRequired[256];
+        return __atomic_is_lock_free(sizeof(lockRequired),&lockRequired);
+    }
+  ]])],
+  [AC_MSG_RESULT(-latomic)
+   ATOMICLIB='-latomic'],
+  [AC_MSG_RESULT(no)]
```

I do not like the fact that we are guessing that linker success means "-latomic is needed" while failure means "no library is needed". Both guesses can be wrong. I suspect that this is the best we can (or want to) do. If that is the case, then let's disclose the fact that we are guessing:

```suggestion
  [AC_MSG_RESULT(-latomic does not hurt and may be required)
   ATOMICLIB='-latomic'],
  [AC_MSG_RESULT(unknown)]
```

You probably want to disregard the exact suggestion in this change request and also edit the AC_MSG_CHECKING() text, to make the whole thing look more precise and sound better. For example:

```
checking whether -latomic is a good idea... probably
checking whether -latomic is a good idea... no
```

---

Aye. The AC_SEARCH_LIBS logic design is better - I think use that logic but with our custom test. Try without the library and if that succeeds declare not-needed. Otherwise try the same test with the library to see if needed. Fail completely if both do.

---

> to make the whole thing look more precise and sound better. For example:

I applied this suggestion.

> I think use that logic but with our custom test

I do not mind keeping to this design - but I could not find a way how to implement it by means of AC_LINK_IFELSE without an extra output noise. In this case, we would need two such macros (for w/o library tests) and each of them would produce a 'checking whether ...' line. AC_SEARCH_LIBS implemented this logic internally, looping through the list of libraries (including an empty string). 
On the other hand, the outcome when the existing AC_LINK_IFELSE results in 'not-needed' seems to contradict our expectations and understanding of how the proposed C++ code snippet should work. This would indicate that for some unknown/strange reason either all atomic methods (including is_lock_free()) proved to be inlined (and no explicit -latomic is required at all) or there are still some methods (other from is_lock_free()), requiring -latomic. Since it is not evident whether we should skip or use -latomic in this case - I would leave the current logic as is - for simplicity sake.

---

> I could not find a way how to implement it by means of AC_LINK_IFELSE without an extra output noise.

I think it is fine to use two (nested) tests, producing two lines in the ./configure output. I would not necessarily classify those lines as "noise". They should be (slightly) different, of course, to reflect the different nature of each test. For example:

    checking whether linking without -latomic works... no
    checking whether linking with -latomic works... yes

If it is not too difficult, please define the test program code once. Sorry, I do not know how to do that using autoconf macros or m4, but I am pretty sure there is a way to avoid code duplication.

> the outcome when the existing AC_LINK_IFELSE results in 'not-needed' seems to contradict our expectations and understanding of how the proposed C++ code snippet should work

This is a difficult question/issue, and my confidence level is low, but I disagree with your conclusion. AFAICT, the "not needed" case means that

* STL was able to provide an inlined version of the is_lock_free() method for a large structure. Please note that an inlined method does not mean that the atomic for the large structure is lock-free -- the inlined method can still return false. Now, I am not sure inlining is_lock_free() for large structures is standard-compliant. It may not be. There is a GCC [bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54005) about it with a long discussion that seem to imply that inlining is not possible because the compiler must ask libatomic (by design). However, it is not clear to me whether that applies to all compilers and all designs.
* The external version of the method was linked in without Squid explicitly adding a (non-existent) libatomic (e.g., by the user providing some custom libraries or the compiler doing the same by default). 

The existence of the second bullet probably trumps any doubts in the first bullet.

> or there are still some methods (other from is_lock_free()), requiring -latomic

That is indeed possible, but we would have to accommodate them as we learn about them. AFAICT, the only viable alternative for addressing this part of your concern is to link with libatomic whenever linking with libatomic is _possible_ (i.e. does not lead to errors). I think we should try what @yadij has proposed first and resort to this (kind of excessive) simpler alternative only if we discover that we cannot quickly find a small set of atomic methods to test with.

---

@eduard, AC_DEFUN can be used to create a macro to de-duplicate any repeated logic (eg the code snippet). If you are unsure there are examples in acinclude/squid-util.m4

I am okay with either alternative; replicating SEARCH_LIB model or always linking to -latomic when possible. Squid usually builds with the --as-needed flag set, so extra libraries in the compiler line does not mean bloat in the final binary.

---

Done.